[2단계 - 자동차 경주 리팩토링] 후추(주찬민) 미션 제출합니다.

Combi153 opened the pull request:
안녕하세요 현구막!

깃허브 이슈를 간신히 해결하고 다시 리뷰 요청드립니다. 보내주신 리뷰에 대해서는 제 생각을 코멘트로 남겨놨어요.

- [Set과 List에 대한 생각](https://github.com/woowacourse/java-racingcar/pull/496#discussion_r1104106386)
- [boolean 메서드 네이밍](https://github.com/woowacourse/java-racingcar/pull/496#discussion_r1104061952)
- [Random에 대한 테스트](https://github.com/woowacourse/java-racingcar/pull/496#discussion_r1103916019)
- [GameRecorder와 GameRecordManager](https://github.com/woowacourse/java-racingcar/pull/496#discussion_r1104047164)

2단계에서는 특별히 GameRecorder와 GameRecordManager 에 대한 구조를 새로 리팩토링하였는데, 제가 짠 방식에 스스로 확신이 들지 않네요.

궁금한 사항은 아래와 같습니다.
- GameRecorder 객체와 RecordManager 객체를 나누어 짠 방식에 대해 어떻게 생각하시는지 궁금해요.
- 기존에 Cars에서 Set 자료형을 List 자료형으로 변경하였는데 그로인해 변경해야 하는 코드들의 목록이 굉장히 많았습니다. 도메인 객체에서 자료형의 변경으로 인해 타 객체들이 영향을 받는 것도 의존성 설계에 미숙함이 드러난 것으로 봐야 할까요?

Combi153 committed:
fix: printResultGuide 메소드를 수정한다

수정 내용 :
접근 제어자를 private로 바꾼다.
메소드의 순서를 변경한다.

Combi153 committed:
refactor: Cars 객체의 클래스 변수 cars의 자료형을 변경한다

변경 내용 :
Set 자료형을 List로 변경한다.
List 초기화 시 변경 불가능한 List로 초기화한다.

Combi153 committed:
fix: Cars 객체의 변화를 반영한다

Combi153 committed:
refactor: GameRecordManager 객체를 리팩토링한다

변경 내용 :
GameRecordManager 객체의 이름을 RecordManagerOfGameResultOfCar 로 변경한다.
RecordManager 인터페이스를 만들고, RecordManagerOfGameResultOfCar 객체가 구현하도록 한다.

Combi153 committed:
refactor: Car 객체의 생성자를 추가 및 CarFactory의 상수의 위치를 옮긴다.

변경 이유 :
자동차 객체의 DEFAULT_POSITION을 아는 것은 자동차 공장이 아니라, 자동차 그 자체가 맞기 때문이다.

Combi153 committed:
chore: GameResultOfCar 클래스의 패키지 위치를 변경한다

Combi153 committed:
fix: Car 생성자에 validate 메소드를 추가한다

Hyeon9mak commented on src/main/java/racingcar/domain/car/Car.java:
```diff
@@ -17,6 +18,12 @@ public Car(String name, int position) {
+        validate(name);
+        this.name = name;
+        this.position = DEFAULT_POSITION;
+    }
```
(취향) 생성자 체이닝을 이용해보면 어떨까요?

```java
public Car(String name) {
    this(name, DEFAULT_POSITION);
}
```

Combi153 replied:
이 방법은 처음 알게 되었습니다. 앞으로 적극 활용해볼 수 있을 것 같아요! 알려주셔서 감사합니다!!!

Hyeon9mak commented on src/main/java/racingcar/domain/cars/Cars.java:
```diff
@@ -3,18 +3,17 @@
         validate(cars);
-        this.cars = new LinkedHashSet<>(cars);
+        this.cars = Collections.unmodifiableList(cars);
     }
```
생각해보시면 좋을 내용이 있어요. [UnmodifiableList은 만능이 아니다. / 방어적 복사](https://ecsimsw.tistory.com/entry/unmodifiableList%EC%9D%80-%EB%A7%8C%EB%8A%A5%EC%9D%B4-%EC%95%84%EB%8B%88%EB%8B%A4)

Combi153 replied:
unmodifiableList는 불변 측면에서 무적이 아니었군요. 정말 좋은 내용입니다!! 헌데 방어적 복사를 위해 새로운 객체를 계속해서 생성한다면, 특정 상황에서는 객체 생성으로 인한 성능 저하가 생길 수도 있지 않을까 하는 생각이 드네요. 이러한 호기심을 해결하려면 자바의 작동방식과 메모리에 관한 공부를 하고, 그런 상황에서 포기하거나 선택할만한 트레이드 오프는 무엇이 있을까 고민해보면 되겠네요. 기록해두고 나중에 꼭 궁금증을 해결해보겠습니다!

Hyeon9mak reviewed (changes requested):
안녕하세요 후추~ 2단계 고생 많으셨어요. 잘 해결되어서 다행이네요 🙏 

> `GameRecorder` 객체와 `RecordManager` 객체를 나누어 짠 방식에 대해 어떻게 생각하시는지 궁금해요.

1단계 PR 코멘트를 확인해보니 '이런 식으로 확장되겠지.' 상상해보신 것 같아요. 의도는 이해가 되고 구현 방법은 취향이라 큰 이견은 없습니다. 

개인적인 의견을 덧붙이자면 이해는 되지만 공감은 안되는 상태에요. 후추가 확장을 우려하는 상황에 대한 자세한 설명이나, 그 상황에서 확장되는 시나리오를 글, 그림 등으로 더 자세히 표현해서 설득해주시면 좋을 것 같아요. (사실 그렇게 해도 대부분 요구사항이 상상한 것보다 더 복잡하게 확장되거나 전혀 다른 방향으로 리팩터링이 이루어져야 하는 경우도 허다해서 당장은 [YAGNI](https://ko.wikipedia.org/wiki/YAGNI) 로 보이네요 🥲)

> 기존에 `Cars` 에서 `Set` 자료형을 `List` 자료형으로 변경하였는데 그로인해 변경해야 하는 코드들의 목록이 굉장히 많았습니다. 도메인 객체에서 자료형의 변경으로 인해 타 객체들이 영향을 받는 것도 의존성 설계에 미숙함이 드러난 것으로 봐야 할까요?

"학습 차원에서 예상치 못한 부분의 리팩터링을 유도했고, 그래서 변경해야하는 코드 목록이 많았다." 가 맞을 것 같아요 🙏 😅 

'내가 경험해보니까 이게 가장 피곤하고 가장 어려운 유지보수 포인트야. 그래서 미리 이렇게 해놔야 해' 가 설계라고 생각해요. 정말 단순하게 "`Cars` 자료형이 변경될 걸 예측하지 못했다." 라면 미숙한게 맞는데, 실제로 자동차 경주 같은 애플리케이션에서 `Cars` 와 같은 객체의 자료형이 자주 변경되진 않을테니...

현업에서도 코드를 작성할 때 보면 계층별 + 방향(request/response) 별로 모든 DTO 를 만드는 분이 있는가 하면, 그냥 하나의 DTO 로 모든 계층 + 방향을 퉁치는 분도 계세요. 나중에 계층, 방향별로 DTO 자료구조가 변할 땐 DTO 를 미리 분리해두신 분이 유리하겠지만, 그 상황이 오기 전까진 DTO 가 그저 [보일러플레이트 코드](https://www.wisewiredbooks.com/term-dict/common/boilerplate.html)일 뿐이죠. 어쩌면 그 상황이 영원히 안올 수도 있구요.

추가로 궁금하신 점 언제든지 질문 주세요!
마지막까지 화이팅입니다!

Combi153 committed:
fix: Car 클래스의 생성자를 수정한다

Combi153 committed:
refactor: Cars 클래스의 생성자 세부내용을 변경한다

변경 내용:
방어적 복사를 활용해 cars 변수를 초기화한다.

Combi153 committed:
docs: 기능 목록을 수정 및 추가한다

Combi153 committed:
refactor: RecordManager 객체 등을 삭제한다

변경 내용 :
RecordManager 인터페이스와 그 구현 객체를 삭제한다.
RecordManagerOfGameResultOfCar 객체의 기능을 GameRecorder 내부로 옮긴다.

Combi153 committed:
refactor: GameResultOfCar 객체에 값을 비교하는 메서드를 추가한다

Combi153 committed:
refactor: findTopPosition 메서드를 수정한다

Combi153 committed:
refactor: 메서드의 인자에 final 키워드를 추가한다

Combi153 commented:
안녕하세요 현구막 2단계 리팩토링이 늦었어요 죄송해요😥

리뷰해주신 내용에 대해 답변하자면

> 개인적인 의견을 덧붙이자면 이해는 되지만 공감은 안되는 상태에요. 후추가 확장을 우려하는 상황에 대한 자세한 설명이나, 그 상황에서 확장되는 시나리오를 글, 그림 등으로 더 자세히 표현해서 설득해주시면 좋을 것 같아요. (사실 그렇게 해도 대부분 요구사항이 상상한 것보다 더 복잡하게 확장되거나 전혀 다른 방향으로 리팩터링이 이루어져야 하는 경우도 허다해서 당장은 [YAGNI](https://ko.wikipedia.org/wiki/YAGNI) 로 보이네요)

저 스스로도 `GameRecorder와 GameRecordManager 에 대한 구조`에 확신을 가질 수 없었고, 그래서 현구막에게 생각을 여쭤봤어요. 현구막의 리뷰를 보면서 왜 스스로 확신을 가질 수 없었는지 깨닫게 되었네요. 제가 예상한 확장이 실제로 일어날 가능성이 적어보였고, 확장이 일어난다고 해도 그 확장을 대비한 구조라고 생각이 들지 않았어요. `확장을 대비하는 척하는 구조`라고 느꼈습니다. 결국 YANGI이자, 고쳐야할 코드가 되었네요. 한편으로는 스스로 확신이 없는 코드를 커밋한 것에 대해 반성하게 되었네요. 개발자로서 코드에 대한 책임을 지는 것이 마땅한데 프로답지 못한 행동을 한 것 같아 부끄럽습니다😞 앞으로 개발하면서 딜레마에 빠질 일이 많을텐데 그때마다 이런 태도를 갖는 건 좋지 못할 거에요. 왜 그 코드가 최선인지 저 스스로 납득이 되도록 고민하고 생각한 후 선택하는 태도를 갖겠습니다.

제가 확장이 일어날 것이라 예상한 것과 제 코드의 한계에 대해 생각한 것은 다음과 같아요.

현재 `GameRecorder` 객체는 gameRoung와 Cars 객체를 받아 자동차 게임의 결과를 기록해요. 기록은 `GameResultOfCar` 객체에 gameRound, carName, position 과 같은 정보를 담아 이뤄지고요. 시간이 지남에 따라 자동차 게임의 결과 기록 방식이 달라질 수 있다고 생각했어요. 가령 자동차의 운전자, 브랜드, 사용 연료량 등의 정보가 추가적으로 담길 수 있을 거에요. 이러한 확장에 `RecordManager` 인터페이스와 그 구현체가 대응할 수 있으면 좋을 것 같았습니다.

하지만 제가 짠 방식에서는 `RecordManager`는 확장에 대응하지 못해요. 기록 방식이 변화한다면 결국 `GameResultOfCar`가 변화하는 것인데 이것은 다른 모든 코드의 수정을 함께 요구하기 때문입니다. `RecordManager`는 있으나 마나 한 인터페이스가 되는 것이지요. 어차피 다른 것들이 모두 바뀌어서요.

현재는 확장에는 닫혀있긴 하나, 저 스스로 확신하는 코드로 리팩토링하였습니다. RecordManager의 기능을 GameRecorder 내부에 옮겼어요.

확장을 대비하는 시도도 경험과 실력이 필요할 거에요. 시간이 지나면 지난 코드들을 다시 보며 그때에는 확장적인 구조로 리팩토링 해보고 싶습니다.

Hyeon9mak commented on src/main/java/racingcar/domain/carfactory/CarFactory.java:
```diff
@@ -8,15 +8,13 @@
 
-    private List<Car> makeCars(List<String> carNames) {
+    private List<Car> makeCars(final List<String> carNames) {
         return carNames.stream()
-                .map(carName -> new Car(carName, DEFAULT_POSITION))
+                .map(Car::new)
                 .collect(Collectors.toUnmodifiableList());
     }
```
생성자 추가로 객체 생성 코드가 훨씬 간결해졌네요 👍

Hyeon9mak commented on src/main/java/racingcar/domain/record/GameRecorder.java:
```diff
@@ -1,24 +1,36 @@
-    public GameRecorder(Set<GameResultOfCar> records, GameRecordManager gameRecordManager) {
+    public GameRecorder(final List<GameResultOfCar> records) {
         this.records = records;
-        this.gameRecordManager = gameRecordManager;
     }
 
-    public void record(int gameRound, Cars cars) {
-        records.addAll(gameRecordManager.makeGameResultsOfCars(gameRound, cars));
+    public void record(final int gameRound, final Cars cars) {
+        records.addAll(makeResults(gameRound, cars));
     }
 
-    public Set<GameResultOfCar> getRecords() {
+    public List<GameResultOfCar> makeResults(final int gameRound, final Cars cars) {
+        return cars.getCars()
+                .stream()
+                .map(car -> createGameResultOfCar(gameRound, car))
+                .collect(Collectors.toUnmodifiableList());
+    }
+
+    private GameResultOfCar createGameResultOfCar(final int gameRound, final Car car) {
+        return new GameResultOfCar(gameRound, car.getName(), car.getPosition());
+    }
+
+    public List<GameResultOfCar> getRecords() {
         return records;
     }
```
`GameRecordManager` 가 사라짐으로 인해서 복잡성이 크게 낮아졌어요. 설계 복잡도도 유지보수성에 포함되는 것 같아요. 팀 동료 개발자들이 코드를 오래 보고, 모두가 비슷한 이해도를 갖춘 후에 조금 더 유연한 설계에 고민해보는 것도 좋은 방법인거 같아요.

Hyeon9mak reviewed (approved):
안녕하세요 후추~ 자동차 경주 2단계 마지막까지 정말 고생 많으셨습니다!
마지막까지 꼼꼼하게 리팩터링해주셨네요 💯 

> '내가 경험해보니까 이게 가장 피곤하고 가장 어려운 유지보수 포인트야. 그래서 미리 이렇게 해놔야 해' 가 설계라고 생각해요.

앞서 말씀 드렸던 것도 비슷한 맥락이었어요. 상상만으로 변경포인트를 추측하는 것과 실제로 겪어본 경험을 기반으로 추측하는 것은 엄연히 다르니까요. 학습이 주 목적이니까 과한 설계를 해보는 경험도 쌓으신거라 생각해요. 경험이 쌓이면 쌓일수록 유연한 설계를 더욱 빠르게 하실 수 있을 것 같아요. 이것 또한 성장이라고 생각하시면 좋겠어요. 😄 

앞으로도 크루, 코치, 리뷰어분들과 다양한 이야기 나누시면서 후추만의 생각을 넓혀나가시면 좋겠습니다!
남은 미션들도 화이팅입니다! 👍
